### ISSUE TITLE
db.lastID is undefined

### ISSUE THREAD

**Issue by Zvorykin** [fritx/cross-sqlcipher#4]
Code
`db.run('INSERT INTO foo ...', function (err) {
            console.log(err, this.lastID)
          })`
prints `null undefined`

Using Electron 1.8

What am I doing wrong?

**Comment by fritx** [comment]
@ElectroDr3am hey, are you sure you were using `function (err) {}` instead of `err => {}`?

Or have you tried:

```js
stmt = db.prepare('INSERT INTO foo VALUES (?)')
stmt.run(params, err => {
  console.log(err, stmt.lastID)
})
```

I've just tried `cross-sqlcipher` in both node and electron, Mac.

(It would take me some time to test it in Windows.)

Ref: 
- https://github.com/mapbox/node-sqlite3/issues/100
- https://github.com/mapbox/node-sqlite3/issues/622#issuecomment-212518985

**Comment by Zvorykin** [comment]
I'm using Electron and Vue.js and create connection in main process like
that
(for test i've changed to `node-sqlite`)

  sqlite3 = require('sqlite3')
// else
// sqlite3 = require('cross-sqlcipher')

let db = new sqlite3.Database('db.db')
global.DB = db

then get it from render

let remote = window.require('electron').remote
  DB = remote.getGlobal('DB')

Vue.prototype.$db = DB

and using in vue js component

methods: {
  createTicket:function () {
    this.$db.run(`INSERT INTO foo...`,
      {
        //params
      }, function (err) {
        console.log(this.lastID, this)
   // that's the point
      })
      .get('SELECT last_insert_rowid() as id', (err, row) => {
        console.log(row.id)
      })
  }
}

The point I've specified `this` is some global variable, I think it is the
whole App variable.

Maybe you can provide example of better use of node-sqlite?

Thanks.


2017-10-04 7:25 GMT+03:00 Fritz Lin <notifications@github.com>:

**Comment by fritx** [comment]
Oh you are using electron's remote and Vue stuff, I'm not quite sure.
I suggest you use `stmt.lastID` syntax. You can try it instead.
See https://github.com/mapbox/node-sqlite3/issues/622#issuecomment-212518985

**Comment by Zvorykin** [comment]
Nope, I've tryied your advice, but same result. Nevermind, I'm working of
small enought project so I can simply query SELECT last_insert_rowid()
as idafter
every INSERT if I need it. Thanks for reply!

2017-10-04 11:18 GMT+03:00 Fritz Lin <notifications@github.com>:

**Comment by fritx** [comment]
@ElectroDr3am hah.. anyway I have another advice...
Personally I think the logic business (db query) should be placed in the main process,
and you turn off the `webPreferences.nodeIntegration` in renderer processes.

You can send ipc messages between them, and don't have to use `remote.getGlobal`,
it could be the ideal model of an electron app?

**Comment by Zvorykin** [comment]
Thanks for advice, I'll consider refactoring my app using that idea.
I think it's worth it and may be usefull to avoid such troubles like I've
got.

2017-10-05 17:27 GMT+03:00 Fritz Lin <notifications@github.com>:

**Comment by Zvorykin** [comment]
By the way I've tested using sqlite from main process and it works
correctly as described in docs.

**Comment by fritx** [comment]
@ElectroDr3am does it work for you? with cross-sqlcipher

**Comment by Zvorykin** [comment]
@fritx Sorry for offtop but does cross-sqlcipher support Promicess? Maybe you can you provide some example? Thanks
